### PR TITLE
Handle OS errors gracefully

### LINKED ISSUE #1795: Handle OS errors gracefully
### Which components does the task require to be changed? (think hard pls)

snforge

### Description

We often panic when writing/reading a file/creating dir/other OS related operation fail. We should bubble those errors instead (they will be pretty printed in main)

### PR DESCRIPTION
**Pull request by Draggu**
Closes #1795

## Introduced changes

- Handle OS errors gracefully

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Review by Arcticae — APPROVED** [review]

File: crates/cheatnet/src/forking/state.rs [hunk 1]
```diff
@@ -40,14 +41,13 @@ pub struct ForkStateReader {
...
-    #[must_use]
-    pub fn new(url: Url, block_number: BlockNumber, cache_dir: &str) -> Self {
-        ForkStateReader {
-            cache: ForkCache::load_or_new(&url, block_number, cache_dir),
+    pub fn new(url: Url, block_number: BlockNumber, cache_dir: &str) -> Result<Self> {
+        Ok(ForkStateReader {
+            cache: ForkCache::load_or_new(&url, block_number, cache_dir)?,
             client: JsonRpcClient::new(HttpTransport::new(url)),
             block_number,
             runtime: Runtime::new().expect("Could not instantiate Runtime"),
```
Comment (drknzz) [inline crates/cheatnet/src/forking/state.rs:49]: context?

File: crates/forge-runner/src/running.rs [hunk 2]
```diff
@@ -379,6 +379,7 @@ fn get_fork_state_reader(
                 *block_number,
                 workspace_root.join(CACHE_DIR).as_ref(),
             )
+            .unwrap()
```
Comment (drknzz) [inline crates/forge-runner/src/running.rs:382]: It's ok right? (just to double check)

**Review by drknzz — APPROVED** [review]
👍

File: crates/cheatnet/src/forking/state.rs (hunk 1 above)
Comment (MaksymilianDemitraszek) [reply crates/cheatnet/src/forking/state.rs:49]: Can this fail in any way by user input?

File: crates/forge-runner/src/running.rs (hunk 2 above)
Comment (Draggu) [reply crates/forge-runner/src/running.rs:382]: No, it's not. Good catch

File: crates/cheatnet/src/forking/state.rs (hunk 1 above)
Comment (Draggu) [reply crates/cheatnet/src/forking/state.rs:49]: Done

**Comment by Draggu** [comment]
Panics in `ForkCache::save` are not changed due to the fact it's called on drop and there is no way to propagate errors

**Comment by Arcticae** [comment]
> Panics in `ForkCache::save` are not changed due to the fact it's called on drop and there is no way to propagate errors

We can change the logic there if it's required for errors to bubble. It's not necessary to have it on `drop`

**Comment by Draggu** [comment]
> > Panics in `ForkCache::save` are not changed due to the fact it's called on drop and there is no way to propagate errors
> 
> We can change the logic there if it's required for errors to bubble. It's not necessary to have it on `drop`

But then it's easy to forget calling `save()`. Either propagate or check if saved by compiler

**Pull request merged**